src/main/java/org/terasology/metalrenegades/combat/system/EnemySpawnSystem.java

```diff
@@ -0,0 +1,209 @@
+import org.terasology.math.geom.Vector2i;
+import org.terasology.math.geom.Vector3f;
+import org.terasology.math.geom.Vector3i;
+import org.terasology.metalrenegades.combat.component.EnemyGracePeriodComponent;
+import org.terasology.metalrenegades.combat.component.NightEnemyComponent;
+import org.terasology.network.ClientComponent;
+import org.terasology.registry.CoreRegistry;
+import org.terasology.registry.In;
+import org.terasology.utilities.random.FastRandom;
+import org.terasology.world.WorldProvider;
+import org.terasology.world.block.Block;
+import org.terasology.world.block.BlockManager;
+import org.terasology.world.sun.OnDawnEvent;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.function.Function;
+
+@RegisterSystem(RegisterMode.AUTHORITY)
+public class EnemySpawnSystem extends BaseComponentSystem implements UpdateSubscriberSystem {
+
+    @In
+    private EntityManager entityManager;
+
+    @In
+    private WorldProvider worldProvider;
+
+    @In
+    private BlockManager blockManager;
+
+    @In
+    private NightTrackerSystem nightTrackerSystem;
+
+    @In
+    private SettlementEntityManager settlementEntityManager;
+
+    /**
+     * Check blocks at and around the target position and check if it's a valid spawning spot
+     */
+    private Function<Vector3i, Boolean> isValidSpawnPosition;
+
+    private FastRandom random;
+
+    private Block sand;
+    private Block air;
+
+    private int cyclesLeft;
+
+    private Queue<EntityRef> enemyQueue;
+
+    private static final int MAX_ENEMIES = 20;
+
+    private boolean ready;
+
+    @Override
+    public void postBegin() {
+        sand = blockManager.getBlock("CoreAssets:Sand");
+        air = blockManager.getBlock(BlockManager.AIR_ID);
+
+        random = new FastRandom();
+        isValidSpawnPosition = this::isValidSpawnPosition;
+
+        nightTrackerSystem = CoreRegistry.get(NightTrackerSystem.class);
+        enemyQueue = Queues.newLinkedBlockingQueue();
+        ready = true;
+    }
+
+    @Override
+    public void update(float delta) {
+        if (cyclesLeft < 300 || !nightTrackerSystem.isNight() || !ready) {
+            cyclesLeft++;
+            return;
+        }
+        cyclesLeft = 0;
+
+        for (EntityRef client : entityManager.getEntitiesWith(ClientComponent.class)) {
+            ClientComponent clientComponent = client.getComponent(ClientComponent.class);
+            EntityRef character = clientComponent.character;
+
+            if (character.hasComponent(EnemyGracePeriodComponent.class)) {
+                EnemyGracePeriodComponent enemyGracePeriodComponent =
+                        character.getComponent(EnemyGracePeriodComponent.class);
+                enemyGracePeriodComponent.cyclesLeft--;
+
+                if (enemyGracePeriodComponent.cyclesLeft >= 0) {
+                    character.removeComponent(EnemyGracePeriodComponent.class);
```

Comment:
This section confuses me. You are decreasing `enemyGracePeriodComponent.cyclesLeft`, but the condition to remove it is if `enemyGracePeriodComponent.cyclesLeft >=0 `.
Should it maybe be `enemyGracePeriodComponent.cyclesLeft <=0 ` ?

Reply:
Yep, this comparison is done the wrong way, thanks for pointing it out! Because of the new spawning system this component has been removed though, so I guess the problem is fixed anyways 😄